### PR TITLE
[SPARK-8333] [SQL] Spark failed to delete temp directory created by HiveContext

### PR DESCRIPTION
**Pull request by navis**
Shutdown underlying megastore and delete temporary directory when the hive context is closed.

### CODE REVIEW

**Comment by andrewor14** [comment]
ok to test @liancheng @yhuai

**Comment by markhamstra** [comment]
I'd rather something more structured like case classes were used instead of plain tuples.  It is less than immediately apparent what things like `state._2` mean.

File: core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala [hunk 1]
```diff
@@ -121,6 +121,9 @@ case class SparkListenerApplicationStart(
 @DeveloperApi
 case class SparkListenerApplicationEnd(time: Long) extends SparkListenerEvent
 
+@DeveloperApi
+case class SparkListenerContextStop() extends SparkListenerEvent
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala]: I'm not sure why you need this. Why won't `SparkListenerApplicationEnd` work? It's send in `SparkContext.stop()`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientWrapper.scala [hunk 2]
```diff
@@ -401,4 +400,14 @@ private[hive] class ClientWrapper(
         client.dropDatabase(db, true, false, true)
       }
   }
+
+  override def shutdown(): Unit = withHiveState {
+    logInfo("Shutting down hive metastore")
+    if (state._1 == Thread.currentThread()) {
```
Comment (vanzin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientWrapper.scala]: Will this ever be true?

This method is called from the `SparkListener` you're adding, which in a live application is called from the listener bus dispatching thread, which is most probably not the thread that created the `state` object.

File: core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala]: +1 I would prefer to limit the number of new events

**Comment by andrewor14** [comment]
@navis are you actively working on this? Do you have time to address @vanzin's comments?

**Comment by JoshRosen** [comment]
@navis, would you mind addressing the comments in this PR or closing it if you don't plan to work on it soon?

**Comment by andrewor14** [comment]
@navis would you mind closing this patch if you don't intend to work on it? We can always reopen it later if there is more interest.

**Pull request closed without merging**